File: frontend/src/api/traderIntelligence.ts
import { supabase } from '../lib/supabase';
import { callEdgeFunction } from '../lib/edgeFunctions';
import type { TradeDebrief, PersonalPattern, PersonalRule, TraderProfile, TraderProfileSnapshot, DashboardData } from './types';

export type { TradeDebrief, PersonalPattern, PersonalRule, TraderProfile, TraderProfileSnapshot, DashboardData };

export const traderIntelligenceService = {
  dashboard: async (projectId: string): Promise<DashboardData> => {
    const { data: profile } = await supabase.from('trader_profile').select('*').eq('project_id', projectId).maybeSingle();
    const { data: debriefs } = await supabase.from('trade_debrief').select('*').eq('project_id', projectId).order('created_at', { ascending: false }).limit(5);
    const { count: debriefCount } = await supabase.from('trade_debrief').select('*', { count: 'exact', head: true }).eq('project_id', projectId);
    const { count: patternCount } = await supabase.from('personal_pattern').select('*', { count: 'exact', head: true }).eq('project_id', projectId).eq('active', true);
    const { count: ruleCount } = await supabase.from('personal_rule').select('*', { count: 'exact', head: true }).eq('project_id', projectId);
    const { count: approvedRuleCount } = await supabase.from('personal_rule').select('*', { count: 'exact', head: true }).eq('project_id', projectId).eq('status', 'approved');
    return { debrief_count: debriefCount ?? 0, pattern_count: patternCount ?? 0, rule_count: ruleCount ?? 0, approved_rule_count: approvedRuleCount ?? 0, profile: profile as TraderProfile | null, recent_debriefs: (debriefs ?? []) as TradeDebrief[] };
  },

  listDebriefs: async (projectId: string, params?: { skip?: number; limit?: number; trade_id?: string }): Promise<TradeDebrief[]> => {
    let query = supabase.from('trade_debrief').select('*').eq('project_id', projectId).order('created_at', { ascending: false });
    if (params?.trade_id) query = query.eq('trade_id', params.trade_id);
    if (params?.limit) query = query.range(params.skip ?? 0, (params.skip ?? 0) + params.limit - 1);
    const { data, error } = await query;
    if (error) throw error;
    return (data ?? []) as TradeDebrief[];
  },

  getDebrief: async (projectId: string, debriefId: string): Promise<TradeDebrief> => {
    const { data, error } = await supabase.from('trade_debrief').select('*').eq('id', debriefId).eq('project_id', projectId).single();
    if (error) throw error;
    return data as TradeDebrief;
  },

  searchDebriefs: async (projectId: string, q: string, limit?: number): Promise<TradeDebrief[]> => {
    const { data, error } = await supabase.from('trade_debrief').select('*').eq('project_id', projectId).or(`summary.ilike.%${q}%,lessons_learned.ilike.%${q}%`).order('created_at', { ascending: false }).limit(limit ?? 20);
    if (error) throw error;
    return (data ?? []) as TradeDebrief[];
  },

  generateDebrief: (projectId: string, tradeId: string): Promise<{ debrief: TradeDebrief; message: string }> =>
    callEdgeFunction('ai', { operation: 'generate-debrief', project_id: projectId, data: { trade_id: tradeId } }),

  updateDebrief: async (projectId: string, debriefId: string, data: Partial<TradeDebrief>): Promise<TradeDebrief> => {
    const { data: row, error } = await supabase.from('trade_debrief').update(data).eq('id', debriefId).eq('project_id', projectId).select().single();
    if (error) throw error;
    return row as TradeDebrief;
  },

  deleteDebrief: async (projectId: string, debriefId: string): Promise<void> => {
    const { error } = await supabase.from('trade_debrief').delete().eq('id', debriefId).eq('project_id', projectId);
    if (error) throw error;
  },

  listPatterns: async (projectId: string, params?: { skip?: number; limit?: number; category?: string; active?: boolean }): Promise<PersonalPattern[]> => {
    let query = supabase.from('personal_pattern').select('*').eq('project_id', projectId).order('occurrence_count', { ascending: false });
    if (params?.category) query = query.eq('category', params.category);
    if (params?.active !== undefined) query = query.eq('active', params.active);
    if (params?.limit) query = query.range(params.skip ?? 0, (params.skip ?? 0) + params.limit - 1);
    const { data, error } = await query;
    if (error) throw error;
    return (data ?? []) as PersonalPattern[];
  },

  getPattern: async (projectId: string, patternId: string): Promise<PersonalPattern> => {
    const { data, error } = await supabase.from('personal_pattern').select('*').eq('id', patternId).eq('project_id', projectId).single();
    if (error) throw error;
    return data as PersonalPattern;
  },

  detectPatterns: (projectId: string, _limit?: number): Promise<PersonalPattern[]> =>
    callEdgeFunction('ai', { operation: 'detect-patterns', project_id: projectId }),

  updatePattern: async (projectId: string, patternId: string, data: Partial<PersonalPattern>): Promise<PersonalPattern> => {
    const { data: row, error } = await supabase.from('personal_pattern').update(data).eq('id', patternId).eq('project_id', projectId).select().single();
    if (error) throw error;
    return row as PersonalPattern;
  },

  deletePattern: async (projectId: string, patternId: string): Promise<void> => {
    const { error } = await supabase.from('personal_pattern').delete().eq('id', patternId).eq('project_id', projectId);
    if (error) throw error;
  },

  listRules: async (projectId: string, params?: { skip?: number; limit?: number; status?: string; category?: string }): Promise<PersonalRule[]> => {
    let query = supabase.from('personal_rule').select('*').eq('project_id', projectId).order('created_at', { ascending: false });
    if (params?.status) query = query.eq('status', params.status);
    if (params?.category) query = query.eq('category', params.category);
    if (params?.limit) query = query.range(params.skip ?? 0, (params.skip ?? 0) + params.limit - 1);
    const { data, error } = await query;
    if (error) throw error;
    return (data ?? []) as PersonalRule[];
  },

  getRulesForApproval: async (projectId: string, limit?: number): Promise<PersonalRule[]> => {
    const { data, error } = await supabase.from('personal_rule').select('*').eq('project_id', projectId).eq('status', 'draft').order('created_at', { ascending: false }).limit(limit ?? 50);
    if (error) throw error;
    return (data ?? []) as PersonalRule[];
  },

  getRule: async (projectId: string, ruleId: string): Promise<PersonalRule> => {
    const { data, error } = await supabase.from('personal_rule').select('*').eq('id', ruleId).eq('project_id', projectId).single();
    if (error) throw error;
    return data as PersonalRule;
  },

  generateRules: (projectId: string): Promise<{ rules: PersonalRule[]; message: string }> =>
    callEdgeFunction('ai', { operation: 'generate-rules', project_id: projectId }),

  approveRule: async (projectId: string, ruleId: string, notes?: string): Promise<PersonalRule> => {
    const { data: row, error } = await supabase.from('personal_rule').update({ status: 'approved', approved_at: new Date().toISOString(), rejection_reason: notes || null }).eq('id', ruleId).eq('project_id', projectId).select().single();
    if (error) throw error;
    return row as PersonalRule;
  },

  rejectRule: async (projectId: string, ruleId: string, reason: string): Promise<PersonalRule> => {
    const { data: row, error } = await supabase.from('personal_rule').update({ status: 'rejected', rejected_at: new Date().toISOString(), rejection_reason: reason }).eq('id', ruleId).eq('project_id', projectId).select().single();
    if (error) throw error;
    return row as PersonalRule;
  },

  updateRule: async (projectId: string, ruleId: string, data: Partial<PersonalRule>): Promise<PersonalRule> => {
    const { data: row, error } = await supabase.from('personal_rule').update(data).eq('id', ruleId).eq('project_id', projectId).select().single();
    if (error) throw error;
    return row as PersonalRule;
  },

  deleteRule: async (projectId: string, ruleId: string): Promise<void> => {
    const { error } = await supabase.from('personal_rule').delete().eq('id', ruleId).eq('project_id', projectId);
    if (error) throw error;
  },

  getProfile: async (projectId: string): Promise<TraderProfile> => {
    const { data, error } = await supabase.from('trader_profile').select('*').eq('project_id', projectId).maybeSingle();
    if (error) throw error;
    return data as TraderProfile;
  },

  buildProfile: (projectId: string): Promise<TraderProfile> =>
    callEdgeFunction('ai', { operation: 'build-profile', project_id: projectId }),

  getSnapshots: async (projectId: string, limit?: number): Promise<TraderProfileSnapshot[]> => {
    const { data, error } = await supabase.from('trader_profile_snapshot').select('*').eq('project_id', projectId).order('snapshot_date', { ascending: false }).limit(limit ?? 20);
    if (error) throw error;
    return (data ?? []) as TraderProfileSnapshot[];
  },
};

File: backend/src/api/routes/trader_intelligence.py
from uuid import UUID
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from src.api.deps import get_db, get_project_or_404
from src.models.project import Project
from src.schemas.trader_intelligence import (
    TradeDebriefRead, TradeDebriefCreate, TradeDebriefUpdate,
    PersonalPatternRead, PersonalPatternCreate, PersonalPatternUpdate,
    PersonalRuleRead, PersonalRuleCreate, PersonalRuleUpdate,
    GenerateDebriefRequest, GenerateDebriefResponse,
    ApproveRuleRequest, RejectRuleRequest, GenerateRulesResponse,
    TraderProfileRead, TraderProfileSnapshotRead,
    DashboardResponse,
)
from src.crud import trader_intelligence as crud
from src.services.trader_intelligence import (
    generate_debrief_from_trade, detect_personal_patterns,
    generate_proposed_rules, build_or_update_profile,
)

router = APIRouter()


# --- Dashboard ---

@router.get("/dashboard", response_model=DashboardResponse)
def intelligence_dashboard(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    summary = crud.get_profile_summary(db, project_id)
    recent = crud.get_debriefs(db, project_id=project_id, limit=5)
    return DashboardResponse(
        debrief_count=summary["debrief_count"],
        pattern_count=summary["pattern_count"],
        rule_count=summary["rule_count"],
        approved_rule_count=summary["approved_rule_count"],
        profile=summary["profile"],
        recent_debriefs=recent,
    )


# --- Trade Debrief ---

@router.get("/debriefs", response_model=list[TradeDebriefRead])
def read_debriefs(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    trade_id: UUID | None = None,
    db: Session = Depends(get_db),
):
    return crud.get_debriefs(db, project_id=project_id, skip=skip, limit=limit, trade_id=trade_id)


@router.get("/debriefs/search", response_model=list[TradeDebriefRead])
def search_debriefs(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    q: str = Query(..., min_length=1),
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
):
    return crud.search_debriefs(db, project_id=project_id, q=q, limit=limit)


@router.get("/debriefs/{debrief_id}", response_model=TradeDebriefRead)
def read_debrief(
    project_id: UUID,
    debrief_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_debrief(db, debrief_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Debrief not found")
    return db_obj


@router.post("/debriefs/generate", response_model=GenerateDebriefResponse)
def generate_debrief(
    project_id: UUID,
    req: GenerateDebriefRequest,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    try:
        debrief = generate_debrief_from_trade(db, req.trade_id)
        return GenerateDebriefResponse(debrief=debrief, message="Debrief generated successfully")
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))


@router.put("/debriefs/{debrief_id}", response_model=TradeDebriefRead)
def update_debrief(
    project_id: UUID,
    debrief_id: UUID,
    obj_in: TradeDebriefUpdate,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_debrief(db, debrief_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Debrief not found")
    return crud.update_debrief(db, db_obj=db_obj, obj_in=obj_in)


@router.delete("/debriefs/{debrief_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_debrief(
    project_id: UUID,
    debrief_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_debrief(db, debrief_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Debrief not found")
    crud.remove_debrief(db, id=debrief_id)


# --- Personal Patterns ---

@router.get("/patterns", response_model=list[PersonalPatternRead])
def read_patterns(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    category: str | None = None,
    active: bool | None = None,
    db: Session = Depends(get_db),
):
    return crud.get_patterns(db, project_id=project_id, skip=skip, limit=limit, category=category, active=active)


@router.get("/patterns/{pattern_id}", response_model=PersonalPatternRead)
def read_pattern(
    project_id: UUID,
    pattern_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_pattern(db, pattern_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pattern not found")
    return db_obj


@router.post("/patterns/detect", response_model=list[PersonalPatternRead])
def detect_patterns(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    limit: int = Query(50, ge=1, le=500),
    db: Session = Depends(get_db),
):
    return detect_personal_patterns(db, project_id, limit=limit)


@router.put("/patterns/{pattern_id}", response_model=PersonalPatternRead)
def update_pattern(
    project_id: UUID,
    pattern_id: UUID,
    obj_in: PersonalPatternUpdate,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_pattern(db, pattern_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pattern not found")
    return crud.update_pattern(db, db_obj=db_obj, obj_in=obj_in)


@router.delete("/patterns/{pattern_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_pattern(
    project_id: UUID,
    pattern_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_pattern(db, pattern_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Pattern not found")
    crud.remove_pattern(db, id=pattern_id)


# --- Personal Rules ---

@router.get("/rules", response_model=list[PersonalRuleRead])
def read_rules(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    status: str | None = None,
    category: str | None = None,
    db: Session = Depends(get_db),
):
    return crud.get_rules(db, project_id=project_id, skip=skip, limit=limit, status=status, category=category)


@router.get("/rules/for-approval", response_model=list[PersonalRuleRead])
def read_rules_for_approval(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
):
    return crud.get_rules_for_approval(db, project_id, limit=limit)


@router.get("/rules/{rule_id}", response_model=PersonalRuleRead)
def read_rule(
    project_id: UUID,
    rule_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_rule(db, rule_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rule not found")
    return db_obj


@router.post("/rules/generate", response_model=GenerateRulesResponse)
def generate_rules(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    rules = generate_proposed_rules(db, project_id)
    return GenerateRulesResponse(rules=rules, message=f"{len(rules)} rules generated/updated")


@router.post("/rules/{rule_id}/approve", response_model=PersonalRuleRead)
def approve_rule(
    project_id: UUID,
    rule_id: UUID,
    req: ApproveRuleRequest = ApproveRuleRequest(),
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_rule(db, rule_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rule not found")
    if db_obj.status != "draft":
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Rule is already {db_obj.status}")
    from datetime import datetime, timezone
    crud.create_rule_version(
        db, rule_id=db_obj.id, version=db_obj.version,
        title=db_obj.title, description=db_obj.description,
        evidence=db_obj.evidence, change_notes=req.notes or "Approved",
    )
    db_obj.status = "approved"
    db_obj.approved_at = datetime.now(timezone.utc)
    db_obj.version += 1
    db.commit()
    db.refresh(db_obj)
    return db_obj


@router.post("/rules/{rule_id}/reject", response_model=PersonalRuleRead)
def reject_rule(
    project_id: UUID,
    rule_id: UUID,
    req: RejectRuleRequest,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_rule(db, rule_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rule not found")
    if db_obj.status != "draft":
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Rule is already {db_obj.status}")
    from datetime import datetime, timezone
    db_obj.status = "rejected"
    db_obj.rejected_at = datetime.now(timezone.utc)
    db_obj.rejection_reason = req.reason
    db.commit()
    db.refresh(db_obj)
    return db_obj


@router.put("/rules/{rule_id}", response_model=PersonalRuleRead)
def update_rule(
    project_id: UUID,
    rule_id: UUID,
    obj_in: PersonalRuleUpdate,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_rule(db, rule_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rule not found")
    crud.create_rule_version(
        db, rule_id=db_obj.id, version=db_obj.version,
        title=db_obj.title, description=db_obj.description,
        evidence=db_obj.evidence, change_notes="Updated before new version",
    )
    updated = crud.update_rule(db, db_obj=db_obj, obj_in=obj_in)
    updated.version += 1
    db.commit()
    db.refresh(updated)
    return updated


@router.delete("/rules/{rule_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_rule(
    project_id: UUID,
    rule_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    db_obj = crud.get_rule(db, rule_id)
    if not db_obj or db_obj.project_id != project_id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rule not found")
    crud.remove_rule(db, id=rule_id)


# --- Trader Profile ---

@router.get("/profile", response_model=TraderProfileRead)
def read_profile(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    profile = crud.get_profile(db, project_id)
    if not profile:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Profile not found. Build it first.")
    return profile


@router.post("/profile/build", response_model=TraderProfileRead)
def build_profile(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    db: Session = Depends(get_db),
):
    return build_or_update_profile(db, project_id)


@router.get("/profile/snapshots", response_model=list[TraderProfileSnapshotRead])
def read_profile_snapshots(
    project_id: UUID,
    project: Project = Depends(get_project_or_404),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
):
    return crud.get_snapshots(db, project_id, limit=limit)
